### PR TITLE
Show the use of proper gogo/googleapi types with status.

### PR DESCRIPTION
**Pull request by johanbrandhorst**
This avoids the use of `google.golang.org/genproto` package types (except for the mandatory status type).

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -31,23 +34,25 @@ func (b *Backend) AddUser(ctx context.Context, user *pbExample.User) (*types.Emp
...
 					Description: "The first user created must have the role of an ADMIN",
 				},
 			},
-		})
-		if err == nil {
-			return nil, detSt.Err()
 		}
-
-		return nil, st.Err()
+		byts, _ := proto.Marshal(b)
```
Comment (awalterschulze) [inline server/server.go]: Remember to check the error.

File: server/server.go [hunk 2]
```diff
@@ -64,6 +69,29 @@ func (b *Backend) ListUsers(_ *types.Empty, srv pbExample.UserService_ListUsersS
...
+		b := &rpc.PreconditionFailure{
+			Violations: []*rpc.PreconditionFailure_Violation{
+				{
+					Type:        "USER",
+					Subject:     "no users created",
+					Description: "No users have been created",
+				},
+			},
+		}
+		byts, _ := proto.Marshal(b)
```
Comment (awalterschulze) [inline server/server.go]: Remember to check the error

File: server/server.go [hunk 3]
```diff
@@ -5,8 +5,11 @@ import (
 	"sync"
 	"time"
 
+	"github.com/gogo/googleapis/google/rpc"
+	"github.com/gogo/protobuf/proto"
 	"github.com/gogo/protobuf/types"
-	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	pbAny "github.com/golang/protobuf/ptypes/any"
```
Comment (awalterschulze) [inline server/server.go]: Are we still planning to get rid of these pbAny and pbStatus types and replace them with gogo types in future?

File: main.go [hunk 4]
```diff
@@ -9,12 +9,8 @@ import (
 	"net"
 	"net/http"
 	"os"
-	"reflect"
-	"strings"
 
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
-	gogoproto "github.com/gogo/protobuf/proto"
-	golangproto "github.com/golang/protobuf/proto"
```
Comment (awalterschulze) [inline main.go:17]: nice :)

File: status/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+The code is 95% copied from the official gRPC status package, so the gRPC
+License applies.
+
+### Changes
+
+The changes applies include changing the use of the
+`golang/protobuf` packages to `gogo/protobuf`, and changing the
+generated files from `google.golang.org/genproto/googleapis` to
+`github.com/goog/googleapis/`.
```
Comment (awalterschulze) [inline status/README.md]: gogo not goog :)

**Review by awalterschulze — APPROVED** [review]

File: status/README.md (hunk 5 above)
Comment (johanbrandhorst) [reply status/README.md]: Whoops

**Comment by johanbrandhorst** [comment]
I'm going to remove the manual vendor changes and wait with merging this until upstream (https://github.com/grpc/grpc-go/pull/1927) has been merged (or rejected :sob:).

**Comment by awalterschulze** [comment]
Sounds like a plan :)

File: Gopkg.toml [hunk 6]
```diff
@@ -45,13 +49,11 @@ required = [
   name = "github.com/mwitkow/go-proto-validators"
 
 [[constraint]]
+  branch = "patch-1"
   name = "google.golang.org/grpc"
-  version = "1.10.0"
+# version = "1.10.0"#
+  source = "github.com/johanbrandhorst/grpc-go"
```
Comment (johanbrandhorst) [inline Gopkg.toml]: TODO: Revert back once https://github.com/grpc/grpc-go/pull/1927 is merged

File: Gopkg.toml [hunk 7]
```diff
@@ -38,20 +38,26 @@ required = [
...
 [[constraint]]
   branch = "master"
-  name = "github.com/golang/mock"
+  name = "github.com/mwitkow/go-proto-validators"
+
+[[override]]
+  branch = "patch-1"
+  name = "google.golang.org/grpc"
+# version = "1.10.0"#
+  source = "github.com/johanbrandhorst/grpc-go"
```
Comment (johanbrandhorst) [inline Gopkg.toml]: TODO: Revert this once https://github.com/grpc/grpc-go/pull/1927 is merged

File: Gopkg.toml [hunk 8]
```diff
@@ -38,20 +38,26 @@ required = [
...
-  version = "1.10.0"
+  branch = "master"
+  name = "github.com/grpc-ecosystem/go-grpc-middleware"
 
 [[constraint]]
   branch = "master"
-  name = "github.com/golang/mock"
+  name = "github.com/mwitkow/go-proto-validators"
+
+[[override]]
```
Comment (johanbrandhorst) [inline Gopkg.toml]: TODO: Set this back to `[[constraint]]`

**Comment by johanbrandhorst** [comment]
Updated in anticipation of grpc/grpc-go#1927. Will just need setting the Gopkg.toml back then we can merge.

**Comment by awalterschulze** [comment]
Ooo wow.  This looks great.  I hope 1927 goes through.  Great work.

**Comment by johanbrandhorst** [comment]
I've update this with the new gRPC version, and gogo/status, and everything seems to work alright.

**Review by awalterschulze — APPROVED** [review]
Beautiful :)

**Pull request merged**